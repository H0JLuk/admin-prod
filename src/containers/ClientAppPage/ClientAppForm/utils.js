import { notification } from 'antd';
import { addSettings, updateSettingsList } from '../../../api/services/settingsService';
import { designKeysForCheck, SETTINGS_TYPES } from './ClientAppFormConstants';

export const showNotify = (message, isError) => {
    const config = {
        duration: 5,
        placement: 'bottomRight',
    };

    if (isError) {
        notification.error({
            ...config,
            message,
        });
    } else {
        notification.success({
            ...config,
            message,
        });
    }

};

export async function createOrUpdateKey (changedParams) {
    const updateSettings = [];
    const addSettingsArr = [];

    changedParams.forEach(({ type, ...params }) => {
        type === SETTINGS_TYPES.EDIT
            ? updateSettings.push(params)
            : addSettingsArr.push(params);
    });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}

export const checkExistDesignSettings = (settings) => {
    return designKeysForCheck.some(key => !Object.prototype.hasOwnProperty.call(settings, key));
};

export const getImage = (imageName) => {
    return `${process.env.PUBLIC_URL}/images/clientAppDesign/${imageName}.png`;
};


