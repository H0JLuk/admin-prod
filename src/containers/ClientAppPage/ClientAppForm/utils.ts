import { customNotifications } from '@utils/notifications';
import { ArgsProps } from 'antd/lib/notification';
import { addSettings, updateSettingsList } from '@apiServices/settingsService';
import { designKeysForCheck, SETTINGS_TYPES } from './ClientAppFormConstants';
import { ISettingObject } from '@types';

export interface IChangedParam {
    clientAppCode: string;
    key: string;
    type: SETTINGS_TYPES;
    value: string;
}

export type IValueToServer = Omit<Partial<IChangedParam>, 'type'>;

export const showNotify = (message: React.ReactNode, isError?: boolean) => {
    const config: ArgsProps = {
        duration: 5,
        message,
    };

    if (isError) {
        customNotifications.error(config);
    } else {
        customNotifications.success(config);
    }

};

type UpdateAndAddSettings = {
    updateSettings: IValueToServer[];
    addSettingsArr: IValueToServer[];
};

export async function createOrUpdateKey(changedParams: IChangedParam[]) {
    const { updateSettings, addSettingsArr } = changedParams.reduce<UpdateAndAddSettings>((result, { type, ...params }) => {
        const key = type === SETTINGS_TYPES.EDIT ? 'updateSettings' : 'addSettingsArr';

        return {
            ...result,
            [key]: [...result[key], params],
        };
    }, { updateSettings: [], addSettingsArr: [] });

    updateSettings.length && (await updateSettingsList(updateSettings));
    addSettingsArr.length && (await addSettings(addSettingsArr));
}

export function checkExistDesignSettings(settings: ISettingObject) {
    return designKeysForCheck.some(key => !Object.prototype.hasOwnProperty.call(settings, key));
}

export function getImage(imageName: string) {
    return `${process.env.PUBLIC_URL}/images/clientAppDesign/${imageName}.png`;
}


