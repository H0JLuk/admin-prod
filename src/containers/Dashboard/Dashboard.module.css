.page {
    display: flex;
    flex-direction: column;
    height: 100%;
}

.header {
    padding: 16px;
}

.header > h3 {
    margin: 0 22px 16px 0;
    font-family: SBSansText-Regular, sans-serif;
    font-size: 20px;
    font-weight: 600;
    font-stretch: normal;
    font-style: normal;
    line-height: 1.4;
    letter-spacing: -0.38px;
    text-align: left;
    color: #171725;
}

.navigationPanel {
    display: flex;
    flex-flow: row nowrap;
    justify-content: flex-start;
    align-items: center;
}

.button {
    margin-right: 16px !important;
    padding: 7px 16px !important;
    border-radius: 2px !important;
    background-color: #09a552;
    font-family: SBSansText-Medium, sans-serif;
    font-size: 14px;
    font-weight: 500;
    font-stretch: normal;
    font-style: normal;
    outline: none 0;
}

.button:focus {
    box-shadow: none;
}

button:last-child {
    margin-right: 0 !important;
}

.menuFilter {
    display: flex;
    flex-flow: row nowrap;
    align-items: center;
    padding: 5px 10px;
    border: grey solid 1px;
    margin-left: 16px;
}

.menuFilter:hover {
    cursor: pointer;
    border-color: green;
}

.menuFilter > p {
    color: black;
    width: 150px;
    font-family: sans-serif;
    font-size: 16px;
    font-weight: bold;
}

.menuFilter > img {
    margin-left: 30px;
}

.content {
    display: flex;
    flex-flow: row wrap;
    justify-content: space-between;
    align-items: stretch;
    padding: 16px 8px;
    background-color: #f0f2f5;
}

.filtersBlock {
    display: flex;
    flex-flow: row wrap;
    justify-content: flex-start;
    align-items: center;
    margin: 8px 0;
}

.filtersBlock:last-child:after {
    content: '';
    background: none;
    border: none;
    height: 0;
    flex-grow: 100;
}
