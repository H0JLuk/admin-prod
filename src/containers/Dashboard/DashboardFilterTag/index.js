import DashboardFilterTag from './DashboardFilterTag';

export default DashboardFilterTag;
