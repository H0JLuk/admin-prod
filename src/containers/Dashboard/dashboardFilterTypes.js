export const WITHOUT_FILTER = '0';
export const BY_APP = '1';
export const BY_DZO = '2';
