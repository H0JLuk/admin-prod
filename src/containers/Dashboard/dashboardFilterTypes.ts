export enum DASHBOARD_FILTERS {
    WITHOUT_FILTER = '0',
    BY_APP = '1',
    BY_DZO = '2',
}
