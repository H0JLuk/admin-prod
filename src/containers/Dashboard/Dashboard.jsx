import React, { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';
import cn from 'classnames';
import { Menu, Dropdown } from 'antd';
import { getDashboardInfo } from '../../api/services/adminService';
import { getClientAppList } from '../../api/services/clientAppService';
import { getDzoList } from '../../api/services/dzoService';
import { getAppCode, saveAppCode } from '../../api/services/sessionService';
import Button from '../../components/Button/Button';
import { goBack, goPromoCampaigns, goToClientApps } from '../../utils/appNavigation';
import DashboardFilterTag from './DashboardFilterTag';
import { WITHOUT_FILTER, BY_APP, BY_DZO } from './dashboardFilterTypes';
import DashboardItem from './DashboardItem';
import downArrowImage from '../../static/svgs/arrow-down.svg';
import styles from './Dashboard.module.css';

const DEFAULT_FILTER_NAME = 'Фильтровать';
const RELOAD_BUTTON_LABEL = 'Обновить';
const CLOSE_BUTTON_LABEL = 'Закрыть';

const DashboardFilterMenu = ({ onClick }) => (
    <Menu onClick={ onClick }>
        <Menu.Item key={ BY_APP }>По приложению</Menu.Item>
        <Menu.Item key={ BY_DZO }>По ДЗО</Menu.Item>
        <Menu.Item key={ WITHOUT_FILTER }>Сбросить</Menu.Item>
    </Menu>
);

const DashboardFilterList = ({ list, filterIdList, onClick }) => {
    return (
        <div className={ styles.filtersBlock }>
            {list.map((elem, index) => (
                <DashboardFilterTag
                    key={ index }
                    { ...elem }
                    handleClick={ onClick }
                    selected={ filterIdList.includes(elem.id) }
                />
            ))}
        </div>
    );
};

const Dashboard = () => {
    const history = useHistory();
    const [filterType, setFilterType] = useState(WITHOUT_FILTER);
    const [filterIdList, setFilterIdList] = useState([]);
    const [filterName, setFilterName] = useState(DEFAULT_FILTER_NAME);
    const [list, setList] = useState([]);
    const [filteredList, setFilteredList] = useState([]);
    const [filterTagList, setFilterTagList] = useState([]);
    const [appList, setAppList] = useState([]);
    const [dzoList, setDzoList] = useState([]);
    const [loading, setLoading] = useState(false);

    const loadData = async () => {
        setLoading(true);
        try {
            const response = await getDashboardInfo();
            setList(response);
            const { clientApplicationDtoList: clientAppList = [] } = await getClientAppList() ?? {};
            setAppList(clientAppList.map(({ id, name }) => ({ id, name })));
            const { dzoDtoList = [] } = await getDzoList() ?? {};
            setDzoList(dzoDtoList.map(({ dzoId: id, dzoName: name }) => ({ id, name })));
        } catch (e) {
            console.error(e?.message);
        } finally {
            setLoading(false);
        }
    };

    useEffect(() => {
        loadData();
    }, []);

    useEffect(() => {
        switch (filterType) {
            case BY_APP:
                setFilterName('Приложение');
                setFilterTagList(appList);
                break;
            case BY_DZO:
                setFilterName('ДЗО');
                setFilterTagList(dzoList);
                break;
            case WITHOUT_FILTER:
            default:
                setFilterName(DEFAULT_FILTER_NAME);
                setFilterTagList([]);
        }
        setFilterIdList([]);
    }, [filterType]);

    useEffect(() => {
        switch (filterType) {
            case BY_APP:
                setFilteredList(list.filter((elem) => filterIdList.includes(elem?.clientApplicationId)));
                break;
            case BY_DZO:
                setFilteredList(list.filter((elem) => filterIdList.includes(elem?.dzoId)));
                break;
            case WITHOUT_FILTER:
            default:
                setFilteredList(list);
        }
    }, [list, filterType, dzoList, appList, filterIdList]);

    const onCloseClick = () => getAppCode() ? goBack(history) : goToClientApps(history);

    // TODO: сделать переход на конкретную промо-кампанию после редизайна
    const onItemClick = (appCode, promoCampaignId) => {
        // console.log({ currentAppCode: getAppCode(), newAppCode: appCode, promoCampaignId });
        saveAppCode(appCode);
        goPromoCampaigns(history);
    };

    const onMenuItemClick = (e) => setFilterType(e?.key);

    const onFilterTagClick = (id, selected) =>
        setFilterIdList(selected ? [...filterIdList, id] : filterIdList.filter((elem) => elem !== id));

    return (
        <div className={ styles.page }>
            <div className={ styles.header } >
                <h3>Дашборд</h3>
                <div className={ styles.navigationPanel }>
                    <Button type="green"
                            label={ RELOAD_BUTTON_LABEL }
                            className={ styles.button }
                            onClick={ loadData }
                            disabled={ loading }
                    />
                    <Dropdown trigger={ ['click'] } overlay={ <DashboardFilterMenu onClick={ onMenuItemClick } /> }>
                        <div className={ styles.menuFilter }>
                            <p>{filterName}</p>
                            <img src={ downArrowImage } alt="" />
                        </div>
                    </Dropdown>
                    <Button type="blue"
                            label={ CLOSE_BUTTON_LABEL }
                            className={ cn(styles.button, styles.exit) }
                            onClick={ onCloseClick }
                    />
                </div>
                <DashboardFilterList list={ filterTagList } filterIdList={ filterIdList } onClick={ onFilterTagClick } />
            </div>
            <div className={ styles.content }>
                {filteredList.map((elem, index) => (
                    <DashboardItem key={ index } handleClick={ onItemClick } { ...elem } />
                ))}
            </div>
        </div>
    );
};

export default Dashboard;
