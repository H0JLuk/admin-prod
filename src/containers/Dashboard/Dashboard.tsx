import { RedoOutlined } from '@ant-design/icons';
import React, { useEffect, useState } from 'react';
import { generatePath, useHistory } from 'react-router-dom';
import { Menu, Dropdown, Button, Alert, MenuProps } from 'antd';
import { getDashboardInfo } from '@apiServices/adminService';
import { getActiveClientApps } from '@apiServices/clientAppService';
import { getDzoList } from '@apiServices/dzoService';
import { getPromoCampaignById } from '@apiServices/promoCampaignService';
import { saveAppCode } from '@apiServices/sessionService';
import useBodyClassForSidebar from '@hooks/useBodyClassForSidebar';
import Header from '@components/Header';
import { requestsWithMinWait } from '@utils/utils';
import DashboardFilterTag from './DashboardFilterTag';
import { DASHBOARD_FILTERS } from './dashboardFilterTypes';
import DashboardItem from './DashboardItem';
import { getPathForPromoCampaignInfo } from '@utils/appNavigation';
import { ClientAppDto, PromoCampaignReport } from '@types';
import downArrowImage from '@svgs/arrow-down.svg';

import styles from './Dashboard.module.css';

interface IFilterList {
    id: number;
    displayName: string;
}

interface IDashboardFilterList {
    list: IFilterList[];
    filterIdList: number[];
    onClick: (id: number, selected: boolean) => void;
}

interface IDashboardFilterMenu {
    onClick: MenuProps['onClick'];
}

type IFilterTagItem = Pick<ClientAppDto, 'id' | 'displayName'>;

const DEFAULT_FILTER_NAME = 'Фильтровать';
const RELOAD_BUTTON_LABEL = 'Обновить';

const EMPTY_DASHBOARD = {
    MESSAGE: 'Отличная работа!',
    DESCRIPTION: 'У нас все под контролем.',
};

const { WITHOUT_FILTER, BY_APP, BY_DZO } = DASHBOARD_FILTERS;

const menuObj = [
    { label: BY_APP, value: 'По приложению' },
    { label: BY_DZO, value: 'По ДЗО' },
    { label: WITHOUT_FILTER, value: 'По умолчанию' },
];


const DashboardFilterMenu: React.FC<IDashboardFilterMenu> = ({ onClick }) => (
    <Menu onClick={onClick}>
        {menuObj.map(({ label, value }) => (
            <Menu.Item key={label}>{value}</Menu.Item>
        ))}
    </Menu>
);

const DashboardFilterList: React.FC<IDashboardFilterList> = ({ list, filterIdList, onClick }) => (
    <div className={styles.filtersBlock}>
        {list.map((elem, index) => (
            <DashboardFilterTag
                key={index}
                {...elem}
                handleClick={onClick}
                selected={filterIdList.includes(elem.id)}
            />
        ))}
    </div>
);

const Dashboard = () => {
    const history = useHistory();
    const [filterType, setFilterType] = useState<DASHBOARD_FILTERS>(WITHOUT_FILTER);
    const [filterIdList, setFilterIdList] = useState<number[]>([]);
    const [filterName, setFilterName] = useState(DEFAULT_FILTER_NAME);
    const [list, setList] = useState<PromoCampaignReport[]>([]);
    const [filteredList, setFilteredList] = useState<PromoCampaignReport[]>([]);
    const [filterTagList, setFilterTagList] = useState<IFilterTagItem[]>([]);
    const [appList, setAppList] = useState<IFilterTagItem[]>([]);
    const [dzoList, setDzoList] = useState<IFilterTagItem[]>([]);
    const [loadingPage, setLoadingPage] = useState(true);
    const [loading, setLoading] = useState(false);

    useBodyClassForSidebar();

    const loadData = async () => {
        setLoading(true);
        try {
            const requests = Promise.all([getDashboardInfo(), getActiveClientApps(), getDzoList()]);
            const [
                dashboardInfo,
                clientAppList,
                { dzoDtoList = [] },
            ] = await requestsWithMinWait(requests);

            setList(dashboardInfo);
            setAppList(clientAppList.map(({ id, displayName }) => ({ id, displayName })));
            setDzoList(dzoDtoList.map(({ dzoId: id, dzoName: displayName }) => ({ id, displayName })));
        } catch (e) {
            console.error(e?.message);
        }
        setLoading(false);
    };

    useEffect(() => {
        (async () => {
            await loadData();
            setLoadingPage(false);
        })();
    }, []);

    useEffect(() => {
        switch (filterType) {
            case BY_APP:
                setFilterName('Приложение');
                setFilterTagList(appList);
                break;
            case BY_DZO:
                setFilterName('ДЗО');
                setFilterTagList(dzoList);
                break;
            case WITHOUT_FILTER:
            default:
                setFilterName(DEFAULT_FILTER_NAME);
                setFilterTagList([]);
        }
        setFilterIdList([]);
    }, [appList, dzoList, filterType]);

    useEffect(() => {
        switch (filterType) {
            case BY_APP:
                setFilteredList(
                    filterIdList.length
                        ? list.filter(({ clientApplicationId: id }) => filterIdList.includes(id))
                        : list,
                );
                break;
            case BY_DZO:
                setFilteredList(
                    filterIdList.length
                        ? list.filter(({ dzoId }) => filterIdList.includes(dzoId))
                        : list,
                );
                break;
            case WITHOUT_FILTER:
            default:
                setFilteredList(list);
        }
    }, [list, filterType, filterIdList]);

    const onItemClick = async (appCode: string, promoCampaignId: number) => {
        saveAppCode(appCode);
        try {
            const path = generatePath(`${ getPathForPromoCampaignInfo() }`, { promoCampaignId });
            const { promoCampaignDtoList = [] } = await getPromoCampaignById(promoCampaignId) ?? {};
            const promoCampaign = promoCampaignDtoList[0];
            if (promoCampaign) {
                history.push(path, { promoCampaign });
            }
        } catch (e) {
            console.error(e?.message);
        }
    };

    const onMenuItemClick: MenuProps['onClick'] = (e) => setFilterType(e.key as DASHBOARD_FILTERS);

    const onFilterTagClick = (id: number, selected: boolean) =>
        setFilterIdList(selected ? [...filterIdList, id] : filterIdList.filter((elem) => elem !== id));

    return (
        <div className={styles.page}>
            <Header buttonBack={false} />
            <div className={styles.header} >
                <h3>Дашборд</h3>
                <div className={styles.navigationPanel}>
                    <Button
                        type="primary"
                        onClick={loadData}
                        loading={loading}
                        icon={<RedoOutlined />}
                    >
                        {RELOAD_BUTTON_LABEL}
                    </Button>
                    <Dropdown
                        trigger={['click']}
                        overlay={<DashboardFilterMenu onClick={onMenuItemClick} />}
                    >
                        <div className={styles.menuFilter}>
                            <p>{filterName}</p>
                            <img src={downArrowImage} alt="" />
                        </div>
                    </Dropdown>
                </div>
                <DashboardFilterList
                    list={filterTagList}
                    filterIdList={filterIdList}
                    onClick={onFilterTagClick}
                />
            </div>
            {!loadingPage && (
                filteredList.length ? (
                    <div className={styles.content}>
                        {filteredList.map((elem, index) => (
                            <DashboardItem key={index} handleClick={onItemClick} {...elem} />
                        ))}
                    </div>
                ) : (
                    <Alert
                        className={styles.message}
                        message={EMPTY_DASHBOARD.MESSAGE}
                        description={EMPTY_DASHBOARD.DESCRIPTION}
                        type="success"
                        showIcon
                    />
                )
            )}
        </div>
    );
};

export default Dashboard;
