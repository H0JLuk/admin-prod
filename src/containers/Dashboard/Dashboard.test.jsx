import React from 'react';
import { shallow, mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import Dashboard from './Dashboard';

import { clientAppListTestResponse, dzoListTestData, promoCampaignTestData } from '../../../__tests__/constants';
import { sleep } from '../../setupTests';
import * as utils from '../../utils/utils';
import * as adminService from '../../api/services/adminService';
import * as clientAppService from '../../api/services/clientAppService';
import * as dzoService from '../../api/services/dzoService';
import * as appNavigation from '../../utils/appNavigation';
import * as promoCampaignService from '../../api/services/promoCampaignService';

const dashboardInfo = [
    {
        active: true,
        clientApplicationCode: 'test',
        clientApplicationDisplayName: 'test name',
        clientApplicationId: 1,
        deleted: true,
        dzoId: 2,
        dzoName: 'string',
        expireInDays: 10,
        issuedCodePercents: 50,
        noCodes: true,
        promoCampaignId: 3,
        promoCampaignName: 'test promo-campaign'
    },
    {
        active: true,
        clientApplicationCode: 'test2',
        clientApplicationDisplayName: 'test name2',
        clientApplicationId: 2,
        deleted: true,
        dzoId: 3,
        dzoName: 'string2',
        expireInDays: 15,
        issuedCodePercents: 80,
        noCodes: true,
        promoCampaignId: 8,
        promoCampaignName: 'test promo-campaign2'
    }
];

const mockHistoryPush = jest.fn();
jest.mock('react-router-dom', () => ({
    useHistory: () => ({
        push: mockHistoryPush,
    }),
    generatePath: jest.fn()
}));

appNavigation.getPathForPromoCampaignInfo = jest.fn();

describe('<Dashboard /> test', () => {
    const container = shallow(<Dashboard />);

    beforeEach(() => {
        adminService.getDashboardInfo = jest.fn(() => dashboardInfo);
        clientAppService.getClientAppList = jest.fn(() => clientAppListTestResponse);
        dzoService.getDzoList = jest.fn(() => ({ dzoDtoList: dzoListTestData }));
    });

    it('should be mount snap', () => {
        expect(container.html()).toMatchSnapshot();
    });

    it('filterName should be "Приложение"', async () => {
        const container = mount(<Dashboard />);
        expect(container.find('p').text()).toEqual('Фильтровать');
        const dropDownContainer = shallow(container.find('Dropdown').at(1).prop('overlay'));
        await act(async () => {
            dropDownContainer.find('Menu').simulate('click', { key: '1' });
        });
        container.update();
        expect(container.find('p').text()).toEqual('Приложение');
    });

    it('filterName should be "ДЗО"', async () => {
        const container = mount(<Dashboard />);
        const dropDownContainer = shallow(container.find('Dropdown').at(1).prop('overlay'));
        await act(async () => {
            dropDownContainer.find('Menu').simulate('click', { key: '2' });
        });
        container.update();
        expect(container.find('p').at(0).text()).toEqual('ДЗО');
    });

    it('onFilterTagClick should be called with selected', () => {
        const container = shallow(<Dashboard />);
        expect(container.find('DashboardFilterList').prop('filterIdList')).toHaveLength(0);
        container.find('DashboardFilterList').simulate('click', 7, true);
        expect(container.find('DashboardFilterList').prop('filterIdList')[0]).toEqual(7);
    });

    it('onFilterTagClick should be called', () => {
        container.find('DashboardFilterList').simulate('click', 7, true);
        container.find('DashboardFilterList').simulate('click', 8, true);
        container.find('DashboardFilterList').simulate('click', 9, true);
        container.find('DashboardFilterList').simulate('click', 7, false);
        expect(container.find('DashboardFilterList').prop('filterIdList')[0]).toEqual(8);
        expect(container.find('DashboardFilterList').prop('filterIdList')).toHaveLength(2);
    });

    it('filterTagList should be correct', async () => {
        utils.requestWithMinWait = jest.fn(() => [adminService.getDashboardInfo(), clientAppService.getClientAppList(), dzoService.getDzoList()]);
        const container = mount(<Dashboard />);
        const dropDownContainer = shallow(container.find('Dropdown').at(1).prop('overlay'));
        await act(async () => {
            dropDownContainer.find('Menu').simulate('click', { key: '2' });
        });
        container.update();
        expect(container.find('DashboardFilterList').prop('list')).toEqual([
            {
                id: 0,
                displayName: 'Беру!'
            },
            {
                id: 3,
                displayName: 'Сбермаркет'
            },
            {
                id: 70,
                displayName: 'СберМобайл'
            },
        ]);
    });

    it('list should be filtered by dzo', async () => {
        utils.requestWithMinWait = jest.fn(() => [adminService.getDashboardInfo(), clientAppService.getClientAppList(), dzoService.getDzoList()]);
        const container = mount(<Dashboard />);
        const dropDownContainer = shallow(container.find('Dropdown').at(1).prop('overlay'));
        await act(async () => {
            dropDownContainer.find('Menu').simulate('click', { key: '2' });
        });
        await act(async () => {
            container.find('DashboardFilterList').prop('onClick')(2, true);
        });
        container.update();
        expect(container.find('DashboardItem')).toHaveLength(1);
        expect(container.find('DashboardItem').prop('dzoId')).toEqual(2);
    });

    it('list should be filtered by application', async () => {
        utils.requestWithMinWait = jest.fn(() => [adminService.getDashboardInfo(), clientAppService.getClientAppList(), dzoService.getDzoList()]);
        const container = mount(<Dashboard />);
        const dropDownContainer = shallow(container.find('Dropdown').at(1).prop('overlay'));
        await act(async () => {
            dropDownContainer.find('Menu').simulate('click', { key: '1' });
        });
        await act(async () => {
            container.find('DashboardFilterList').prop('onClick')(1, true);
        });
        container.update();
        expect(container.find('DashboardItem')).toHaveLength(1);
        expect(container.find('DashboardItem').prop('clientApplicationId')).toEqual(1);
    });

    it('requestWithMinWait should be called', async () => {
        utils.requestWithMinWait = jest.fn(() => [adminService.getDashboardInfo(), clientAppService.getClientAppList(), dzoService.getDzoList()]);
        await act(async () => {
            mount(<Dashboard />);
        });
        expect(utils.requestWithMinWait).toHaveBeenCalledTimes(1);
    });

    it('console.error should be called with "test error" message', async () => {
        const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => '');
        utils.requestWithMinWait = jest.fn(() => Promise.reject(new Error ('test error')));
        await act(async () => {
            mount(<Dashboard />);
        });
        expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
        expect(consoleErrorSpy).toHaveBeenCalledWith('test error');
    });

    it('getPromoCampaignById should be called', async () => {
        utils.requestWithMinWait = jest.fn(() => [adminService.getDashboardInfo(), clientAppService.getClientAppList(), dzoService.getDzoList()]);
        promoCampaignService.getPromoCampaignById = jest.fn(() => ({ promoCampaignDtoList: [promoCampaignTestData] }));
        const container = mount(<Dashboard />);
        await sleep();
        container.update();
        await act(async () => {
            container.find('.item').at(0).simulate('click');
        });
        expect(promoCampaignService.getPromoCampaignById).toHaveBeenCalledTimes(1);
        expect(mockHistoryPush).toHaveBeenCalled();
    });

    it('history.push should not be called', async () => {
        utils.requestWithMinWait = jest.fn(() => [adminService.getDashboardInfo(), clientAppService.getClientAppList(), dzoService.getDzoList()]);
        promoCampaignService.getPromoCampaignById = jest.fn();
        const container = mount(<Dashboard />);
        await sleep();
        container.update();
        await act(async () => {
            container.find('.item').at(0).simulate('click');
        });
        expect(mockHistoryPush).not.toHaveBeenCalled();
    });

    it('console.error should be called, when there is an error ', async () => {
        utils.requestWithMinWait = jest.fn(() => [adminService.getDashboardInfo(), clientAppService.getClientAppList(), dzoService.getDzoList()]);
        promoCampaignService.getPromoCampaignById = jest.fn(() => Promise.reject(new Error ('test error')));
        const consoleErrorSpy = jest.spyOn(console, 'error').mockImplementation(() => '');
        const container = mount(<Dashboard />);
        await sleep();
        container.update();
        await act(async () => {
            container.find('.item').at(0).simulate('click');
        });
        expect(consoleErrorSpy).toHaveBeenCalledTimes(1);
        expect(consoleErrorSpy).toHaveBeenCalledWith('test error');
    });
});
