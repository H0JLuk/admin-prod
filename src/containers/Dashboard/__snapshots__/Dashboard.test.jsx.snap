// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> test should be mount snap 1`] = `"<div class=\\"page\\"><div class=\\"container\\"><div class=\\"avatar\\"><span role=\\"img\\" aria-label=\\"user\\" style=\\"font-size:22px;color:#FFF;background-color:#c1c8d1;border-radius:50px;width:32px;height:32px;padding:3px\\" tabindex=\\"-1\\" class=\\"anticon anticon-user ant-dropdown-trigger dropdown\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"user\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M858.5 763.6a374 374 0 00-80.6-119.5 375.63 375.63 0 00-119.5-80.6c-.4-.2-.8-.3-1.2-.5C719.5 518 760 444.7 760 362c0-137-111-248-248-248S264 225 264 362c0 82.7 40.5 156 102.8 201.1-.4.2-.8.3-1.2.5-44.8 18.9-85 46-119.5 80.6a375.63 375.63 0 00-80.6 119.5A371.7 371.7 0 00136 901.8a8 8 0 008 8.2h60c4.4 0 7.9-3.5 8-7.8 2-77.2 33-149.5 87.8-204.3 56.7-56.7 132-87.9 212.2-87.9s155.5 31.2 212.2 87.9C779 752.7 810 825 812 902.2c.1 4.4 3.6 7.8 8 7.8h60a8 8 0 008-8.2c-1-47.8-10.9-94.3-29.5-138.2zM512 534c-45.9 0-89.1-17.9-121.6-50.4S340 407.9 340 362c0-45.9 17.9-89.1 50.4-121.6S466.1 190 512 190s89.1 17.9 121.6 50.4S684 316.1 684 362c0 45.9-17.9 89.1-50.4 121.6S557.9 534 512 534z\\"></path></svg></span></div></div><div class=\\"header\\"><h3>Дашборд</h3><div class=\\"navigationPanel\\"><button type=\\"button\\" class=\\"ant-btn ant-btn-primary\\"><span role=\\"img\\" aria-label=\\"redo\\" class=\\"anticon anticon-redo\\"><svg viewBox=\\"64 64 896 896\\" focusable=\\"false\\" data-icon=\\"redo\\" width=\\"1em\\" height=\\"1em\\" fill=\\"currentColor\\" aria-hidden=\\"true\\"><path d=\\"M758.2 839.1C851.8 765.9 912 651.9 912 523.9 912 303 733.5 124.3 512.6 124 291.4 123.7 112 302.8 112 523.9c0 125.2 57.5 236.9 147.6 310.2 3.5 2.8 8.6 2.2 11.4-1.3l39.4-50.5c2.7-3.4 2.1-8.3-1.2-11.1-8.1-6.6-15.9-13.7-23.4-21.2a318.64 318.64 0 01-68.6-101.7C200.4 609 192 567.1 192 523.9s8.4-85.1 25.1-124.5c16.1-38.1 39.2-72.3 68.6-101.7 29.4-29.4 63.6-52.5 101.7-68.6C426.9 212.4 468.8 204 512 204s85.1 8.4 124.5 25.1c38.1 16.1 72.3 39.2 101.7 68.6 29.4 29.4 52.5 63.6 68.6 101.7 16.7 39.4 25.1 81.3 25.1 124.5s-8.4 85.1-25.1 124.5a318.64 318.64 0 01-68.6 101.7c-9.3 9.3-19.1 18-29.3 26L668.2 724a8 8 0 00-14.1 3l-39.6 162.2c-1.2 5 2.6 9.9 7.7 9.9l167 .8c6.7 0 10.5-7.7 6.3-12.9l-37.3-47.9z\\"></path></svg></span><span>Обновить</span></button><div class=\\"ant-dropdown-trigger menuFilter\\"><p>Фильтровать</p><img src=\\"arrow-down.svg\\" alt=\\"\\"/></div></div><div class=\\"filtersBlock\\"></div></div></div>"`;
